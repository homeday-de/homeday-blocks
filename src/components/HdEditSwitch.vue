<template>
  <section
    :class="{
      'edit-switch': true,
      'edit-switch--editing-allowed': editingAllowed,
      'edit-switch--loading': loading,
    }"
  >
    <div
      v-if="loading"
      class="edit-switch__loading-indicator"
    >
      <span class="edit-switch__loading-indicator__icon"/>
      <span class="edit-switch__loading-indicator__label">
        {{ t.GENERAL.SAVING }}
      </span>
    </div>
    <div
      v-else-if="isEditMode && editingAllowed"
      class="edit-switch__actions"
    >
      <button
        class="btn edit-switch__actions__button edit-switch__actions__button--cancel"
        @click="cancel">
        <HdIcon
          :src="BUTTONS.cancel.icon"
          class="edit-switch__actions__button__icon"
        />
        <p class="edit-switch__actions__button__label">{{ t.GENERAL.CANCEL }}</p>
      </button>
      <button
        class="btn edit-switch__actions__button edit-switch__actions__button--save"
        @click="save">
        <HdIcon
          :src="BUTTONS.save.icon"
          class="edit-switch__actions__button__icon"
        />
        <p class="edit-switch__actions__button__label">{{ t.GENERAL.SAVE }}</p>
      </button>
    </div>
    <div
      v-else-if="!isEditMode && editingAllowed"
      class="edit-switch__actions"
    >
      <button
        class="btn edit-switch__actions__button edit-switch__actions__button--edit"
        @click="edit">
        <HdIcon
          :src="BUTTONS.edit.icon"
          class="edit-switch__actions__button__icon"
        />
        <p class="edit-switch__actions__button__label">{{ t.GENERAL.EDIT }}</p>
      </button>
    </div>
    <div
      v-if="$slots.default"
      ref="defaultSlot"
      class="edit-switch__view"
    >
      <slot/>
    </div>
    <div
      v-if="$slots.edit && isEditMode && editingAllowed"
      ref="editSlot"
      class="edit-switch__edit"
    >
      <slot name="edit"/>
    </div>
    <div
      v-else-if="$slots.view"
      ref="viewSlot"
      class="edit-switch__view"
    >
      <slot name="view"/>
    </div>
  </section>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import {
  editIcon,
  checkIcon,
  closeIcon,
} from 'homeday-blocks/src/assets/small-icons';

const BUTTONS = {
  edit: {
    icon: editIcon,
  },
  save: {
    icon: checkIcon,
  },
  cancel: {
    icon: closeIcon,
  },
};

export default {
  name: 'HdEditSwitch',
  components: {
    HdIcon,
  },
  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    editing: {
      type: Boolean,
      default: false,
    },
    editingAllowed: {
      type: Boolean,
      default: true,
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
  },
  data() {
    return {
      BUTTONS,
      isEditMode: this.editing,
    };
  },
  watch: {
    editing(value) {
      this.isEditMode = value;
    },
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
  },
  methods: {
    save() {
      this.$emit('save');
    },
    edit() {
      this.isEditMode = true;
      this.$emit('edit');
    },
    cancel() {
      this.isEditMode = false;
      this.$emit('cancel');
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.edit-switch {
  $_root: &;

  display: flex;
  flex-direction: column;
  &__actions {
    display: flex;
    justify-content: flex-end;
    margin-bottom: $stack-s;
    &__button {
      display: flex;
      align-items: center;
      padding: 0;
      margin-left: $inline-m;
      cursor: pointer;
      &--edit {
        color: getShade($secondary-color, 110);
      }
      &--save {
        color: $success-color;
      }
      &--cancel {
        color: $error-color;
      }
      &__label {
        font-size: 18px;
        font-weight: 400;
        line-height: 26px;
        margin-left: $inline-xs;
      }
      &__icon path {
        fill: currentColor;
      }
    }
  }

  &__loading-indicator {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    margin-bottom: $stack-s;

    &__icon {
      width: 20px;
      height: 20px;
      border-radius: 50%;
      border: 2px solid getShade($secondary-color, 110);
      border-right-color: transparent;
      animation: edit-switch-loading-indicator-spin 1s linear infinite;
      content: '';
    }

    &__label {
      display: none;
      margin-left: $inline-s;
      font-size: 18px;
      line-height: 26px;

      @media (min-width: $break-tablet) {
        display: block;
      }
    }
  }

  &__view,
  &__edit {
    transition: opacity $time-s ease-in-out;

    #{$_root}--loading & {
      opacity: 0.3;
      pointer-events: none;
    }
  }

  &--editing-allowed .details-table-wrap:first-child {
    padding-top: 0;
    margin-top: $stack-s * -1;
  }

  h3 + .field {
    margin-top: $stack-s;
  }
}

@keyframes edit-switch-loading-indicator-spin {
  from { transform: rotateZ(0deg); }
  to { transform: rotateZ(360deg); }
}
</style>
