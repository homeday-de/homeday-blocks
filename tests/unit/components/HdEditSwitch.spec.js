import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import { createWrapper } from '@vue/test-utils';
import HdEditSwitch from '@/components/HdEditSwitch.vue';

const DEFAULT_SLOT_CONTENT = '<p>Default slot</p>';
const EDIT_SLOT_CONTENT = '<p>Edit slot</p>';
const VIEW_SLOT_CONTENT = '<p>View slot</p>';
const EDIT_CONTROLS_SELECTOR = '.edit-switch__actions';
const LOADING_INDICATOR_SELECTOR = '.edit-switch__loading-indicator';

const wrapperBuilder = wrapperFactoryBuilder(HdEditSwitch, {
  slots: {
    default: DEFAULT_SLOT_CONTENT,
    edit: EDIT_SLOT_CONTENT,
    view: VIEW_SLOT_CONTENT,
  },
});

describe('HdEditSwitch', () => {
  it('renders the component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Default slot is always visible', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.vm.$refs.defaultSlot).not.toBeUndefined();

    wrapper.setProps({
      editing: true,
    });

    expect(wrapper.vm.$refs.defaultSlot).not.toBeUndefined();
    expect(
      createWrapper(
        wrapper.vm.$refs.defaultSlot,
      ).html().includes(DEFAULT_SLOT_CONTENT),
    ).toBe(true);
  });

  it('In the default state, edit slot is not present, and view is', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.vm.$refs.editSlot).toBeUndefined();
    expect(wrapper.vm.$refs.viewSlot).not.toBeUndefined();
    expect(
      createWrapper(
        wrapper.vm.$refs.viewSlot,
      ).html().includes(VIEW_SLOT_CONTENT),
    ).toBe(true);
  });

  it('In the editing state, edit slot is present, and view is not', () => {
    const wrapper = wrapperBuilder({
      props: {
        editing: true,
      },
    });

    expect(wrapper.vm.$refs.editSlot).not.toBeUndefined();
    expect(
      createWrapper(
        wrapper.vm.$refs.editSlot,
      ).html().includes(EDIT_SLOT_CONTENT),
    ).toBe(true);
    expect(wrapper.vm.$refs.viewSlot).toBeUndefined();
  });

  it('Loading should work', () => {
    const wrapper = wrapperBuilder({
      props: {
        loading: true,
      },
    });

    expect(wrapper.find(LOADING_INDICATOR_SELECTOR).exists()).toBe(true);
  });

  it('You should be able to disallow editing', () => {
    const wrapper = wrapperBuilder({
      props: {
        editingAllowed: false,
      },
    });

    expect(wrapper.find(EDIT_CONTROLS_SELECTOR).exists()).toBe(false);
  });

  it('Should fire edit events', () => {
    const wrapper = wrapperBuilder();

    wrapper.vm.edit();

    expect(wrapper.emitted('edit')).toBeTruthy();

    wrapper.vm.cancel();

    expect(wrapper.emitted('cancel')).toBeTruthy();

    wrapper.vm.save();

    expect(wrapper.emitted('save')).toBeTruthy();
  });
});
