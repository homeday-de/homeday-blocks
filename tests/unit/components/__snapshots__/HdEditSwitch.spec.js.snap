// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdEditSwitch renders the component 1`] = `
<section class="edit-switch edit-switch--editing-allowed">
  <div class="edit-switch__actions"><button class="btn edit-switch__actions__button edit-switch__actions__button--edit">
      <!---->
      <p class="edit-switch__actions__button__label">Bearbeiten</p>
    </button></div>
  <div class="edit-switch__view">
    <p>Default slot</p>
  </div>
  <div class="edit-switch__view">
    <p>View slot</p>
  </div>
</section>
`;
